Fix Admin API clone

bcduggan opened the pull request:
In my testing, I could not clone system templates, like fedora-41-xfce and fedora-41-minimal, from `mgmtvm`:

```yaml
---
- hosts: localhost
  connection: local
  tasks:
      - name: Clone fedora-41-xfce template
        qubesos:
          guest: fedora-41-xfce-clone
          vmtype: "TemplateVM"
          state: present
          template: "fedora-41-xfce"
```

I think this was due to a couple of issues:

1. The [Admin API does not support appmenus](https://github.com/QubesOS/qubes-issues/issues/4809) on Qubes OS 4.2
2. I did not want to add system templates to `include/admin-local-rwx`, which would allow `mgmtvm` to modify their attributes.

This PR...

1. Fixes the first issue by setting the `ignore_errors` flag on `clone_vm` when the local qube's name is not "dom0"
2. Adds a new example policy that allows only the RPCs necessary to clone target qubes to address the second issue
3. Fixes a bug that caused cloned templates to receive the label in the task's parameters instead of the parent's label

Closes QubesOS/qubes-issues#9932

bcduggan committed:
Add unmanaged qube clone example

Add example of cloning an unmanaged qube from a management qube using
the admin API with a minimal policy.

bcduggan committed:
Set ignore_errors on clone_vm when local qube is not dom0

As of QubesOS 4.2, the Admin API doesn't support appmenus, so cloning
qubes from management qubes fails unless the `ignore_errors` flag is
set: https://github.com/QubesOS/qubes-issues/issues/4809

This change sets the `ignore_errors` flag in clone_vm() when the local
qube name is not "dom0".

bcduggan committed:
Only change label of cloned qube for standalone creation

Standalone qube creation is actually a clone operation. Previously, the
qubesos module would update the new qube's label when the guest qube
was either a standalone or template qube.

When the guest is a template qube (vmtype: TemplateVM), the module
shouldn't update the label after cloning.

jmynes commented:
Glad I found this PR, since I was just having this issue as well, whether trying to clone from ansible, or directly via `qvm-clone` in `mgmtvm`

Will test with your patch on my end

jmynes commented:
After applying your patch (I manually just edited the two lines in `/usr/share/ansible/plugins/modules/qubesos.py`, and am otherwise still on the [fepitre](https://github.com/fepitre/qubes-ansible) version), I still get the same errors reported to dom0, _but_, the template does not get deleted, and the script successfully continues and makes the AppVM using our newly cloned TemplateVM:
![error + non deleted](https://github.com/user-attachments/assets/a1046409-f063-40d0-8905-5077ffe9aa88)

Despite the errors, everything goes smoothly:
![error + success](https://github.com/user-attachments/assets/4fd3daf1-7dd8-41ab-8eaa-d7086d09782a)

Here's my matching terminal output:
```
PLAY [local] ************************************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************
ok: [localhost]

TASK [Create fedora-demo] ***********************************************************************************************************************************************************
changed: [localhost]

TASK [Create vault-demo AppVM] ******************************************************************************************************************************************************
changed: [localhost]

TASK [Configure vault-demo properties] **********************************************************************************************************************************************
ok: [localhost]

TASK [Check if vault-demo is running] ***********************************************************************************************************************************************
ok: [localhost]

TASK [Run the vault-demo VM] ********************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP **************************************************************************************************************************************************************************
localhost                  : ok=6    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

jmynes commented:
Basically, the errors are _ignored_, but not suppressed, correct?

bcduggan commented:
Great to hear it worked, eventually! It sounds like you ran into the two issues I ran into, but those denial notifications are annoying. They might also indicate something off in your policies.

> After applying your patch (I manually just edited the two lines in `/usr/share/ansible/plugins/modules/qubesos.py`, and am otherwise still on the [fepitre](https://github.com/fepitre/qubes-ansible) version),

That should be enough. It's a tiny change that works around [this issue](https://github.com/QubesOS/qubes-issues/issues/4809). Notice that on your `mgmtvm`, there's no `qvm-appmenus` tool like there is on `dom0`.

> I still get the same errors reported to dom0

I can't reproduce these with `30-admin-clone` and `include/admin-clone` in the example.

> but, the template does not get deleted, and the script successfully continues and makes the AppVM using our newly cloned TemplateVM

In my testing, it was possible to clone a qube without all the RPCs in the `30-admin-clone.policy`. I added RPCs until I didn't see any denials after cloning. There are a lot of things you can do usefully and still get RPC denials. Try `qvm-ls` from `mgmtvm` and see how many policy denials you get from trying to call RPCs on `dom0`.

> Basically, the errors are _ignored_, but not suppressed, correct?

No RPC executions or errors are ignored or suppressed. If you see a denial notification, `mgmtvm` tried to do something the policy doesn't allow. If you don't see a denial notification, then the application you ran executed any RPC it called successfully.

> [admin@dom0 ~]$ qvm-tags ans-fedora-40-fips
> audiovm-dom0
> created-by-dom0
> created-by-mgmtvm
> guivm-dom0

You shouldn't need to apply the `created-by-mgmtvm` tag manually. That tag should be reserved for qubes `mgmtvm` actually created. Note that `mgmtvm` can change almost anything about qubes with that tag. I wrote `30-admin-clone` in the example to allow only the RPCs necessary to clone a qube without being able to change it.

However, I did notice that when I added only the tag `created-by-mgmtvm` to a test template and removed it from `include/admin-clone`, I got one of the denials you got: "Denied __admin.vm.device.testclass.List__ from __mgmtvm__ to __test_template__". I didn't get the other denial.

I don't get any denials when I add a template to `include/admin-clone` and use the `30-admin-clone` policy. First, make sure this line matches in your `30-admin-clone` policy:
```
!include-service admin.vm.property.Get       * include/admin-clone
```

Then try removing the `created-by-mgmtvm` tag from `ans-fedora-40-fips` and add this line to `include/admin-clone`:
```
mgmtvm ans-fedora-40-fips allow target=dom0
```

How does cloning work with that policy?

bcduggan commented:
Your policies look correct now.

> And after removing created-by-mgmtvm tag from ans-fedora-40-fips, I now get:
> ```
> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: qubesadmin.exc.QubesDaemonAccessError: Service call error: Request refused fatal: [localhost]: FAILED! => {"changed": false, "msg": "Service call error: Request refused\n"}`
> ```

This looks like it could be another RPC denial. Do you see a denial notification when this happens?

Can you clone this template with `qvm-clone` from `mgmtvm`?

Try minimizing your playbook to only the clone task, then try again with `-vvv`. If you still get errors, what RPC denial notifications do you see, if any? What does your minimized playbook contain? What's the verbose output?

marmarek commented:
PipelineRetry

bcduggan commented:
@fepitre For sure, I'll add tests as soon as I can.

jmynes commented:
One other bit of feedback that I just realized, if a `qubes-ansible` task uses this to clone a qube, it starts the created qube, even if that qube already exists. Is this intentional?

marmarek commented:
given #21 is merged, is this one still relevant?
